### PR TITLE
HostCope: Add Customer IP Annotation

### LINKED ISSUE #17026: Will cilium IPAM support specified ip address of pod?
## Proposal / RFE

**Is your feature request related to a problem?**

**Describe the solution you'd like**

As the title states, in some cases we need a static pod ip address because our program does not support DNS.  We think if a pod can be assign with a static ip address will solve this problem. We wonder if cilium will support it in future versions or maybe there're better ways we haven't found? 
We found that calico IPAM has a feature by setting annotation "cni.projectcalico.org/ipAddrs" can specific ip address to single pod, but cilium has more other powerful features that we need, so we decide use cilium as our network plugin.

Looking forward to your favorable  reply.

### PR DESCRIPTION
**Pull request by llhhbc**
Signed-off-by: LongHui Li <llhwonderoflighthbc@gamil.com>

Please ensure your pull request adheres to the following guidelines:

- [ ] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [ ] All code is covered by unit and/or runtime tests where feasible.
- [x] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [x] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [ ] Provide a title or release-note blurb suitable for the release notes.
- [ ] Thanks for contributing!

Fixes:  #17026

kind/feature

### CODE REVIEW

**Comment by llhhbc** [comment]
kind/feature

**Comment by llhhbc** [comment]
I’m not sure if it’s appropriate to add it like this, please submit a version for discussion.

**Comment by twpayne** [comment]
Before diving into the nits of this PR, why is this needed? Please put the explanation in the commit message.

File: pkg/ipam/types.go [hunk 1]
```diff
@@ -12,6 +12,10 @@ import (
 	"github.com/davecgh/go-spew/spew"
 )
 
+const (
+	customPodIpAddr = "cni.cilium.io/ipAddrs"
```
Comment (joestringer) [inline pkg/ipam/types.go:16]: I don't have background on this proposal at all, but I noticed this so I figure I'll mention it: a minor nit, to ensure consistency between different annotations and make it easy to locate all annotations, we manage these through `pkg/annotation/k8s.go`. The format is also a bit different, the prefix should be `io.cilium`.

**Comment by llhhbc** [comment]
> Before diving into the nits of this PR, why is this needed? Please put the explanation in the commit message.

We need to run the database instance in the container. The database has strict ip control. Once the pod’s ip is determined, the subsequent ip should be fixed and bound to the machine (this is achieved by other means)
So we need to have the function of specifying ip

File: pkg/ipam/types.go (hunk 1 above)
Comment (llhhbc) [reply pkg/ipam/types.go:16]: Okay, how do you think this constant should be named

Comment (llhhbc) [reply pkg/ipam/types.go:16]: use  `io.cilium.cni/ipAddrs` ?

**Comment by twpayne** [comment]
> We need to run the database instance in the container. The database has strict ip control. Once the pod’s ip is determined, the subsequent ip should be fixed and bound to the machine (this is achieved by other means)
> So we need to have the function of specifying ip

Have you considered using Cilium's [egress gateway](https://docs.cilium.io/en/v1.10/gettingstarted/egress-gateway/) functionality?

**Comment by llhhbc** [comment]
> > We need to run the database instance in the container. The database has strict ip control. Once the pod’s ip is determined, the subsequent ip should be fixed and bound to the machine (this is achieved by other means)
> > So we need to have the function of specifying ip
> 
> Have you considered using Cilium's [egress gateway](https://docs.cilium.io/en/v1.10/gettingstarted/egress-gateway/) functionality?

This feature is not enough to meet my needs. I am running a database service, which requires the listening address to be a fixed ip address at startup, and at the same time it will broadcast the address to other nodes for other nodes to access the node. This ip requirement is completely determined during initialization, and subsequent changes are not allowed

**Comment by joestringer** [comment]
> This feature is not enough to meet my needs. I am running a database service, which requires the listening address to be a fixed ip address at startup, and at the same time it will broadcast the address to other nodes for other nodes to access the node. This ip requirement is completely determined during initialization, and subsequent changes are not allowed

I'm not sure I follow the reason why it doesn't work yet. When I look at the [Egress NAT Gateway documentation](https://docs.cilium.io/en/v1.10/gettingstarted/egress-gateway/), the [CRD](https://docs.cilium.io/en/v1.10/gettingstarted/egress-gateway/#create-egress-nat-policy) can be created even before your application starts.

File: pkg/ipam/types.go (hunk 1 above)
Comment (joestringer) [reply pkg/ipam/types.go:16]: Sorry, I jumped ahead a bit and I think that we should explore @twpayne's questions before we come to a conclusion on this. The naming will depend on the semantics you are expecting from the feature but I don't yet have my head wrapped around the overall problem or how this PR fits into the solution.

**Comment by llhhbc** [comment]
> > This feature is not enough to meet my needs. I am running a database service, which requires the listening address to be a fixed ip address at startup, and at the same time it will broadcast the address to other nodes for other nodes to access the node. This ip requirement is completely determined during initialization, and subsequent changes are not allowed
> 
> I'm not sure I follow the reason why it doesn't work yet. When I look at the [Egress NAT Gateway documentation](https://docs.cilium.io/en/v1.10/gettingstarted/egress-gateway/), the [CRD](https://docs.cilium.io/en/v1.10/gettingstarted/egress-gateway/#create-egress-nat-policy) can be created even before your application starts, with an IP address that you provide.

This function is very powerful, but it can only manage export addresses. The main reason for this is that the database itself requires a lot of special requirements. For example, etcd, the address it requires to monitor, and the address that it broadcasts to other nodes, all need to be strongly bound to information such as certificates. There are three parts involved: 1. The listening address seen in netstat 2. The external can only use this address and port to access 3. The client also needs to use this address as the source address to access other services. The main reason is that the database has strict requirements, and there are very few modifications that can be provided, only according to its requirements.

I understand that this function can only meet the requirements of 1 and 3, but the requirement of 2 cannot be satisfied.

**Comment by llhhbc** [comment]
@joestringer Is there any progress on this issue? Can it be advanced?

**Comment by twpayne** [comment]
Thanks for this PR. We've been thinking about this and related use cases and are currently working on a more general solution that we hope to propose soon.

File: pkg/ipam/hostscope.go [hunk 2]
```diff
@@ -52,11 +58,29 @@ func (h *hostScopeAllocator) Release(ip net.IP) error {
...
+			return nil, fmt.Errorf("get pod %s info failed %v. ", owner, err)
+		}
+		if pod.Annotations != nil && pod.Annotations[customPodIpAddr] != "" {
+			ip := net.ParseIP(pod.Annotations[customPodIpAddr])
+			if ip == nil {
+				return nil, fmt.Errorf("customer invalid ip: %s. ", pod.Annotations[customPodIpAddr])
+			}
+			err = h.allocator.Allocate(ip)
+			if err != nil {
+				return nil, fmt.Errorf("customer ip is not avaliable %s. ", ip.String())
```
Comment (kvaps) [inline pkg/ipam/hostscope.go:74]:
```suggestion
				return nil, fmt.Errorf("customer ip is not avaliable %s: %w", ip.String(), err)
```

Comment (kvaps) [inline pkg/ipam/hostscope.go:74]:
```suggestion
                               return nil, fmt.Errorf("customer ip is not avaliable %s: %w", ip.String(), err)
```

Comment (llhhbc) [reply pkg/ipam/hostscope.go:74]: Can this mr be merged? @kvaps

**Comment by llhhbc** [comment]
https://github.com/cilium/cilium/issues/17026

**Pull request closed without merging**